### PR TITLE
Add new APIs to azure-json

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Adds new APIs to `azure-json` types.

- `JsonOptions` can now be configured to allow for non-conformant JSON to be parsed (comments, trailing commas, and unescaped control characters). This was added to help ease `azure-identity`'s migration to `azure-json` as it requires parsing of JSON files that may have non-conformant JSON (particularly comments).
- `JsonReader` added `getRawText` which returns the raw JSON string for the current token. For most token types this has no effect but for field names and string values this will return the JSON with escaped characters, such as escaped quotes (`\"`) or encoded specialty characters.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by lmolkova — APPROVED** [review]

File: sdk/core/azure-json-gson/src/main/java/com/azure/json/gson/GsonJsonReader.java [hunk 1]
```diff
@@ -88,17 +91,25 @@ static JsonReader fromReader(Reader json, JsonOptions options) {
...
     private GsonJsonReader(Reader reader, boolean resetSupported, byte[] jsonBytes, String jsonString,
         JsonOptions options) {
-        this(reader, resetSupported, jsonBytes, jsonString, options.isNonNumericNumbersSupported());
+        this(reader, resetSupported, jsonBytes, jsonString, options.isNonNumericNumbersSupported(),
+            options.isAllowComments(), options.isAllowTrailingCommas(), options.isAllowUnescapedControlCharacters());
     }
 
     private GsonJsonReader(Reader reader, boolean resetSupported, byte[] jsonBytes, String jsonString,
-        boolean nonNumericNumbersSupported) {
+        boolean nonNumericNumbersSupported, boolean allowComments, boolean allowTrailingCommas,
```
Comment (conniey) [inline sdk/core/azure-json-gson/src/main/java/com/azure/json/gson/GsonJsonReader.java:99]: Are there tests for these? Or am I missing them

Comment (alzimmermsft) [reply sdk/core/azure-json-gson/src/main/java/com/azure/json/gson/GsonJsonReader.java:99]: I've added tests for comments as it's the most common case:

https://github.com/Azure/azure-sdk-for-java/pull/35698/files#diff-5699bbbfb83540e5e58f85c70107a344477fdcb6efe75f51250a91ea66bdc5b4R833

I can add more tests for trailing commas, not too sure what would be the best way to handled unescaped control characters.

File: sdk/core/azure-json-gson/src/main/java/com/azure/json/gson/GsonJsonProvider.java [hunk 2]
```diff
@@ -25,31 +31,49 @@ public GsonJsonProvider() {
...
     @Override
     public JsonWriter createWriter(Writer json, JsonOptions options) {
+        validate(json, options, JSON_WRITER_EXCEPTION);
+
         return GsonJsonWriter.toWriter(json, options);
     }
+
+    private static void validate(Object json, JsonOptions options, String exceptionMessage) {
+        if (json == null || options == null) {
+            throw new NullPointerException(exceptionMessage);
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/core/azure-json-gson/src/main/java/com/azure/json/gson/GsonJsonProvider.java:76]: I'll handle that in a follow up

File: sdk/core/azure-json-gson/src/main/java/com/azure/json/gson/GsonJsonProvider.java [hunk 3]
```diff
@@ -17,6 +17,12 @@
  * Implementation of {@link JsonProvider} that creates instances using GSON.
  */
 public class GsonJsonProvider implements JsonProvider {
+    private static final String JSON_READER_EXCEPTION = "Both 'json' and 'options' must be passed as non-null to "
+        + "create an instance of JsonReader.";
+
+    private static final String JSON_WRITER_EXCEPTION = "Both 'json' and 'options' must be passed as non-null to "
+        + "create an instance of JsonWriter.";
```
Comment (JonathanGiles) [inline sdk/core/azure-json-gson/src/main/java/com/azure/json/gson/GsonJsonProvider.java:24]: You require non-null here, but in the implementation you have code that allows for `options` to be null. Is there an inconsistency here?

Comment (alzimmermsft) [reply sdk/core/azure-json-gson/src/main/java/com/azure/json/gson/GsonJsonProvider.java:24]: Yes, but the implementation can only be called by these APIs or other implementation code called by these APIs, meaning non-nullness should already be guaranteed. If it's a concern, I can add more null checks.

**Pull request closed without merging**